vm/devices/storage/disklayer_sqlite/src/lib.rs

```diff
@@ -0,0 +1,617 @@
+
+        Ok(SqliteDiskLayer {
+            conn: Arc::new(Mutex::new(conn)),
+            meta,
+            read_only,
+        })
+    }
+
+    async fn write_maybe_overwrite(
+        &self,
+        buffers: &RequestBuffers<'_>,
+        sector: u64,
+        overwrite: bool,
+    ) -> Result<(), DiskError> {
+        assert!(!self.read_only);
+
+        let count = buffers.len() / self.meta.sector_size as usize;
+        tracing::trace!(sector, count, "write");
+
+        let buf = buffers.reader().read_all()?;
+        unblock({
+            let conn = self.conn.clone();
+            let sector_size = self.meta.sector_size;
+            move || write_sectors(conn, sector_size, sector, buf, overwrite)
+        })
+        .await
+        .map_err(|e| DiskError::Io(std::io::Error::new(std::io::ErrorKind::Other, e)))?;
+
+        Ok(())
+    }
+}
+
+impl LayerAttach for FormatOnAttachSqliteDiskLayer {
+    type Error = anyhow::Error;
+    type Layer = SqliteDiskLayer;
+
+    async fn attach(
+        self,
+        lower_layer_metadata: Option<disk_layered::DiskLayerMetadata>,
+    ) -> Result<Self::Layer, Self::Error> {
+        let len = {
+            let lower_len = lower_layer_metadata
+                .as_ref()
+                .map(|m| m.sector_count * m.sector_size as u64);
+            self.format_dbhd
+                .len
+                .or(lower_len)
+                .context("no base layer to infer sector_count from")?
+        };
+        // FUTURE: make sector-size configurable
+        let sector_size = lower_layer_metadata.map(|x| x.sector_size).unwrap_or(512);
+
+        SqliteDiskLayer::new(
+            self.dbhd_path,
+            self.read_only,
+            Some(FormatParams {
+                logically_read_only: self.format_dbhd.logically_read_only,
+                len,
+                sector_size,
+            }),
+        )
+    }
+}
+
+impl LayerIo for SqliteDiskLayer {
+    fn layer_type(&self) -> &str {
+        "sqlite"
+    }
+
+    fn sector_count(&self) -> u64 {
+        self.meta.sector_count
+    }
+
+    fn sector_size(&self) -> u32 {
+        self.meta.sector_size
+    }
+
+    fn is_read_only(&self) -> bool {
+        self.read_only
+    }
+
+    fn disk_id(&self) -> Option<[u8; 16]> {
+        None
+    }
+
+    fn physical_sector_size(&self) -> u32 {
+        self.meta.sector_size
+    }
+
+    fn is_fua_respected(&self) -> bool {
+        true
+    }
+
+    async fn read(
+        &self,
+        buffers: &RequestBuffers<'_>,
+        sector: u64,
+        mut marker: SectorMarker<'_>,
+    ) -> Result<(), DiskError> {
+        let sector_count = (buffers.len() / self.meta.sector_size as usize) as u64;
+        let end_sector = sector + sector_count;
+        tracing::trace!(sector, sector_count, "read");
+        if end_sector > self.meta.sector_count {
+            return Err(DiskError::IllegalBlock);
+        }
+
+        let valid_sectors = unblock({
+            let conn = self.conn.clone();
```

Comment:
Seems like we might want a connection pool to support concurrent IOs (assuming sqlite supports this to a useful degree).

Reply:
Agreed, I think that'd be a more elegant solution here, vs. using a single shared connections.

What's the correct way to properly "size" the pool? Is there a hard-cap on the number of concurrent IOs we might expect? If its unbounded - what should the behavior be when the pool is being fully utilized? Do we have any existing abstractions for this sort of thing?

Haven't done too much involved work in the storage space, so any/all context here would be good

* * * 

But also, as discussed offline - we might want to punt on this for now, as I'm going on vacation tomorrow, so it'd be great to get an MVP in before I head out.